### PR TITLE
[None][test] GVR top-K decode UT: compile-signature-aware parametrization (27m31s -> 13m34s)

### PR DESCRIPTION
**Pull request by longcheng-nv**
## Dev Engineer Review

- Replaced Cartesian parametrization with compile-signature-aware covering matrices.
- Reduced redundant compile-axis combinations while preserving required cross-coverage.
- Kept runtime-only axes parametrized to preserve test IDs, `-k` selection, and granular failure reporting.
- Updated decode, load-balance, R0-equivalence, plateau, and launch-autoconfiguration tests.
- Restricted secant-versus-R0 comparisons to `fp32`.
- Used tie-aware index-set equality for launch-autoconfiguration checks.
- Added targeted `fp16` plateau coverage for `rank_scatter_cs1` and `base_r0off`.
- Kept `test_lb_prepare_partition` unchanged.
- No public API, configuration, infrastructure, or test-list changes.
- Reported B200 results decreased from 828 to 200 passing cases and from 27:31 to approximately 14:20.
- Review should confirm that all required compile signatures remain covered.

## QA Engineer Review

- Modified test file: `tests/unittest/_torch/attention/sparse/test_cute_dsl_gvr_topk_decode.py`.
- Modified existing decode, R0-equivalence, load-balance, plateau, and launch-autoconfiguration tests.
- No test functions were added or removed.
- Compile-axis covering matrices reduce redundant test cases.
- Runtime-only axes remain parametrized.
- Plateau coverage includes `fp16` `rank_scatter_cs1` and `base_r0off` cases.
- No `tests/integration/test_lists/` coverage changes were reported.
- CI and manual-QA coverage for the modified test functions is not confirmed.
- Verdict: needs follow-up.

### Problem

This file is the second longest unittest entry in the repo: 33.61 min in
`tests/integration/defs/.test_durations`, behind only `unittest/_torch/thop/parallel`.
It is scheduled in four stages (`l0_b200`, `l0_b300`, `l0_dgx_b300`,
`l0_gb300_multi_gpus`) and runs serially, because `agg_unit_mem_df.csv` has no
B200/B300 row.

### Why it is slow

The cost is cuTe DSL codegen, not the number of cases. With `--durations=0` on a
B200, the main sweep makes 384 calls and only 84 of them take 5 ms or more. The
first case of each kernel variant pays 3-35 s of JIT; every later case with the same
variant costs under 5 ms. 165 distinct variants x 9.99 s = 1649 s, against 1651 s of
measured wall clock.

The runner keys its compile cache on `(dtype, top_k, next_n, compress_ratio,
cluster_size, pick_tuning(dtype, num_rows, N // cluster_size), seqlen_sorted)`. That
splits the parameters in two:

| | parameters | cost of one more value |
|---|---|---|
| compile axes | `dtype`/`top_k`, `next_n`, `compress_ratio`, `cluster_size`, `N` | one codegen, ~10 s |
| runtime axes | `varlen`, `batch_size`, `preidx_hit_rate`, `hint` | free, reuses a variant |

The sweeps took a full Cartesian product over both kinds, so most of the wall clock
went into recompiling variants the compile axes had already covered.

### What this PR does

**1. Compile axes: covering design instead of a full product.**

| test | compiles before | after |
|---|---:|---:|
| `test_cute_dsl_gvr_topk_decode` | 64 | 16 |
| `..._r0_equivalence` | 40 | 18 |
| `test_lb_main_branches` + `test_lb_vs_reference` | 24 | 12 |
| `..._plateau_terminal` | 10 | 7 |
| `..._launch_autoconfig` | 8 | 6 |

In the main sweep `(dtype, top_k) x (next_n, cr)` stays fully crossed, because
`N_eff = (seq_len - next_n + nn + 1) // cr` is the only real 2-way interaction in the
index math. `(N, cluster_size)` is then laid out as a Latin square over those 16
cells, which makes `(dtype, top_k) x (N, cs)` and `(next_n, cr) x (N, cs)` fully
crossed as well, 16/16 pairs each. Only 3-way and higher interactions are dropped.

**2. Runtime axes stay in `parametrize`.**

They are free serially, because the JIT cache lives in the process and is
reused across cases: 860.01 s with them parametrized against 800.24 s with them
folded into a loop, part of which is run-to-run noise. An earlier revision of
this PR folded them; @yuxianq and @limin2021 pushed back, and they are right for
the way this case runs today, so the folding is reverted and the per-combination
test ids, `-k` selectability and granular failure reporting are kept.

The folding was aimed at pytest-xdist, where each worker is a separate process
with its own cache and `--dist load` scatters one compile cell's combinations
across workers. At `-n 8` that measured 2.52x aggregate codegen inflation with
`parametrize` versus 1.36x with a loop, 344.6 s against 221.7 s wall. That is
not how this case runs today, and if it is ever enabled, `xdist_group` plus
`--dist loadgroup` gives the same locality without giving up `parametrize`.

### Result

Serially on one idle B200:

| | collected | passed | skipped | wall clock |
|---|---:|---:|---:|---:|
| before | 828 | 684 | 144 | 1651.49 s (27:31) |
| after | 200 | 200 | 0 | 809.83 s (13:29) |

The 144 collect-then-skip cases are gone: no cell in the new design is
degenerate.

CI agrees. In `L0_MergeRequest_PR` #54592 on B300, all 200 cases of this file
passed with 16.3 min of summed case time, against the 33.61 min this entry
currently carries in `tests/integration/defs/.test_durations`.

Should `agg_unit_mem_df.csv` ever gain a `parallel_factor` row for these GPUs,
the same file drops to 344.6 s (5:44) at `-n 8`. That file is infra-owned and
wants a memory measurement first, so it is left for a separate PR.

### The one coverage cut: LB

The LB kernel runs only when `counters` is passed (`lb_mode = counters is not None`).
The only production call site of `trtllm::cute_dsl_gvr_topk_decode` is
`indexer.py:1817`, and it passes `order_row` but never `counters` /
`max_batch_size`; `cute_dsl_gvr_topk_lb_prepare` has no caller outside this test
file. So no production shape reaches LB today, while its tests were 24 of the file's
165 compiles at 15-20 s each, the most expensive per compile in the file.

`order_row` has two meanings, which is easy to misread: without `counters` it is the
LJF single-CTA sort hint, and that one is in production; with `counters` it is the LB
partition, and that one is not.

LB keeps `test_lb_prepare_partition` unchanged, all three branch scenarios with
`mixed_half` on both dtypes, the `(next_n x cr)` cross on fp32, both SMEM layouts,
and the full `(varlen x preidx_hit_rate)` cross at every retained cell. Only
compile-axis combinations were dropped, no runtime behaviour. Logical checks go
182 -> 65. Note that none of the 182 pre-existing cases ever passed `cluster_size`,
so LB `cs=2` / `cs=8` were untested before this PR and still are.

@limin2021 you wrote the LB kernel in #15304 — could you confirm that path is still
unwired? If it gets wired later, that PR should thicken this grid again.

### Note for #16953

It raises this entry to `TIMEOUT (120)` in `l0_b300.yml` / `l0_dgx_b300.yml`, which
can probably come back down on rebase. It also adds the `--ignore` split to
`l0_b200.yml`, which this PR does not duplicate.

## Test Coverage

Tests only. The file itself is the coverage: 104 cases, green twice on a B200
(serial and `-n 8`). `pre-commit run --files <this file>` passes, including `ruff`,
`ruff-format` and "Validate test list entries exist in source files (AST)".

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Review by limin2021 — APPROVED** [review]

File: tests/unittest/_torch/attention/sparse/test_cute_dsl_gvr_topk_decode.py [hunk 1]
```diff
@@ -254,15 +303,48 @@ def test_cute_dsl_gvr_topk_decode(
...
     The LJF host-side dispatch order (``order_row``) is covered by the
     dedicated ``test_cute_dsl_gvr_topk_decode_seqlen_sorted`` below on
     representative cells instead of doubling this whole sweep.
     """
     if N - next_n + 1 < top_k:
         pytest.skip(f"N_eff < top_k ({N - next_n + 1} < {top_k}) is a degenerate path")
-    if varlen and batch_size < 2:
-        pytest.skip("varlen with batch_size<2 collapses to fixed")
 
+    for varlen, batch_size, preidx_hit_rate in _DECODE_RUNTIME_CELLS:
```
Comment (yuxianq) [inline tests/unittest/_torch/attention/sparse/test_cute_dsl_gvr_topk_decode.py]: Can we still parametrize runtime cells? I think JIT cache can be reused across different test cases.

File: tests/unittest/_torch/attention/sparse/test_cute_dsl_gvr_topk_decode.py [hunk 2]
```diff
@@ -1079,13 +1231,18 @@ def test_cute_dsl_gvr_topk_decode_launch_autoconfig(dtype, top_k, N, batch_size,
...
+    # extra compiled kernel per cell.
+    if dtype != torch.float32:
+        return
     out_sec = torch.empty(num_rows, top_k, dtype=torch.int32, device="cuda")
     _GvrTopKKernel.launch(logits, pre_idx, seq_lens, out_sec, top_k, enable_r0=False)
     torch.cuda.synchronize()
     _gvr_check(tie_aware_check, out_sec, logits, seq_lens, top_k, next_n=1, compress_ratio=1)
-    if dtype == torch.float32:
-        assert torch.equal(out.sort(dim=-1).values, out_sec.sort(dim=-1).values)
+    assert torch.equal(out.sort(dim=-1).values, out_sec.sort(dim=-1).values)
```
[the remark this answers is not in the document]
Comment (longcheng-nv) [reply tests/unittest/_torch/attention/sparse/test_cute_dsl_gvr_topk_decode.py]: Fixed in 98b9f12 — swapped to `_assert_index_sets_equal_tie_aware(out, out_sec, logits)` and reworded the comment above it (it wrongly claimed fp32 is tie-free).

For the record, the assert is pre-existing (upstream line 1088; this PR only moved the fp32 gate to an early `return`), but the finding is correct and the helper was already in this module, so it is fixed here rather than left as a latent flake.

`test_cute_dsl_gvr_topk_decode_launch_autoconfig`: 4 passed on a B200.

(Apologies for the first version of this reply — the code spans got eaten before posting.)

File: tests/unittest/_torch/attention/sparse/test_cute_dsl_gvr_topk_decode.py (hunk 1 above)
Comment (limin2021) [reply tests/unittest/_torch/attention/sparse/test_cute_dsl_gvr_topk_decode.py]: yes. it will reuse. I think moving them to parameter will be better too.

Comment (longcheng-nv) [reply tests/unittest/_torch/attention/sparse/test_cute_dsl_gvr_topk_decode.py]: Reverted in 352b57e — the runtime axes are back on `parametrize`, and `_runtime_cell()` is dropped with them.

You are right for the way this file runs in CI today. Measured serially on a B200:

| runtime axes | wall clock |
|---|---:|
| `parametrize` | 860.01 s (14:20) |
| loop | 800.24 s (13:20) |

60 s, part of which is run-to-run noise. The 27:31 -> 14:20 win comes from the compile-axis covering design, not from the folding, so it is not worth giving up per-combination test ids, `-k` selectability and granular failure reporting.

For the record on why I folded them in the first place: the JIT cache is per process, and under `pytest-xdist` each worker has its own. `--dist load` hands out one case at a time, so the four runtime combinations of a single compile cell can land on four different workers and each re-runs the same cuTe DSL codegen. At `-n 8`:

| runtime axes | aggregate codegen | wall clock |
|---|---:|---:|
| `parametrize` | 2162.0 s | 344.6 s |
| loop | 1166.9 s | 221.7 s |

That only matters once `agg_unit_mem_df.csv` gains a `parallel_factor` row for these GPUs, which it does not have today. When it does, `@pytest.mark.xdist_group` + `--dist loadgroup` gives the same locality without giving up `parametrize`, which is the better fix anyway. Left for that change.

200 passed on a B200 at 352b57e.

**Review by yuxianq — APPROVED** [review]

File: tests/unittest/_torch/attention/sparse/test_cute_dsl_gvr_topk_decode.py (hunk 1 above)
Comment (yuxianq) [reply tests/unittest/_torch/attention/sparse/test_cute_dsl_gvr_topk_decode.py]: Thanks~
